src/flint/types/fmpq_mpoly.pyx

```diff
@@ -0,0 +1,600 @@
+
+            >>> ctx = fmpq_mpoly_ctx.get_context(4, "lex", 'x')
+            >>> ctx.nvars()
+            4
+        """
+        return self.val.zctx.minfo.nvars
+
+    cpdef ordering(self):
+        """
+        Return the term order of the context object.
+
+            >>> ctx = fmpq_mpoly_ctx.get_context(4, "deglex", 'w')
+            >>> ctx.ordering()
+            'deglex'
+        """
+        if self.val.zctx.minfo.ord == ordering_t.ORD_LEX:
+            return "lex"
+        if self.val.zctx.minfo.ord == ordering_t.ORD_DEGLEX:
+            return "deglex"
+        if self.val.zctx.minfo.ord == ordering_t.ORD_DEGREVLEX:
+            return "degrevlex"
+
+    def gen(self, slong i):
+        """
+        Return the `i`th generator of the polynomial ring
+
+            >>> ctx = fmpq_mpoly_ctx.get_context(3, 'degrevlex', 'z')
+            >>> ctx.gen(1)
+            z1
+        """
+        cdef fmpq_mpoly res
+        assert i >= 0 and i < self.val.zctx.minfo.nvars
+        res = fmpq_mpoly.__new__(fmpq_mpoly)
+        res.ctx = self
+        fmpq_mpoly_init(res.val, res.ctx.val)
+        res._init = True
+        fmpq_mpoly_gen(res.val, i, res.ctx.val)
+        return res
+
+    def constant(self, z):
+        """
+        Create a constant polynomial in this context
+        """
+        cdef fmpq_mpoly res
+        z = any_as_fmpq(z)
+        if z is NotImplemented:
+            raise ValueError("A constant fmpq_mpoly is a fmpq")
+        res = fmpq_mpoly.__new__(fmpq_mpoly)
+        res.ctx = self
+        fmpq_mpoly_init(res.val, res.ctx.val)
+        res._init = True
+        fmpq_mpoly_set_fmpq(res.val, (<fmpq>z).val, res.ctx.val)
+        return res
+
+    def fmpq_mpoly_from_dict(self, d):
+        """
+        Create a fmpz_mpoly from a dictionary.
+
+        The dictionary's keys are tuples of ints (or anything that implicitly converts
+        to fmpz) representing exponents, and corresponding coefficient values of fmpq.
+
+            >>> ctx = fmpq_mpoly_ctx.get_context(2,'lex','x,y')
+            >>> ctx.fmpq_mpoly_from_dict({(1,0):2, (1,1):3, (0,1):1})
+            3*x*y + 2*x + y
+        """
+        cdef long n
+        cdef fmpq_t coefficient
+        cdef int xtype
+        cdef fmpz_struct *exponents
+        cdef fmpz_struct **exp_ptr
+        cdef int nvars = self.nvars()
+        cdef int i,j
+        cdef int count
+        cdef fmpq_mpoly res
+
+        if not PyDict_Check(d):
+            raise ValueError("expected a dictionary")
+        n = PyDict_Size(d)
+        exponents = <fmpz_struct *> libc.stdlib.calloc(nvars, sizeof(fmpz_struct))
+        if exponents == NULL:
+            raise MemoryError()
+        exp_ptr = <fmpz_struct **> libc.stdlib.calloc(nvars, sizeof(fmpz_struct *))
+        if exp_ptr == NULL:
+            libc.stdlib.free(exponents)
+            raise MemoryError()
+        for i in range(nvars):
+            fmpz_init(exponents + i)
+            exp_ptr[i] = exponents + i
```

Comment:
This logic would be better moved to the superclass `flint_mpoly` and shared by all subclasses.

This applies to many things...

Reply:
I completely agree, however when sharing a method, Cython cannot determine the proper C types of some variables and casting becomes a pain point. Particularly when we need access to the underlying flint struct for some C call. I'm sure it could be however I haven't attempted it yet

Reply:
I would imagine something like:
```cython
cdef class flint_mpoly:
    cdef from_dict(...):
        # lots of stuff
        return self._from_dict(...)

cdef class flint_fmpq_mpoly(flint_mpoly):
    cdef _from_dict(...):
        # call C functions here
```
Then all shared logic goes to the superclass but the specific C function calls are spelled out in the subclass as needed.

Reply:
Ah ok I see what you mean. I'm sure that can be done. This code also seems a little fishy to me, I'll take a closer look today and sort out a nicer method, the use of `fmpz_clear` and construction of exponent vectors doesn't seem right to me when we have a perfectly good `fmpz` object. I can understand if this method is supposed to be "faster" however I'd like to substantiate that before hand.

Also bare in mind this comment is on the `fmpq_mpoly` method, this isn't a file I have touched yet and it will receive more  same attention once the `fmpz_mpoly` class is in a model worthy state.